File: app/src/main/java/com/sysu/edu/academic/RegistrationActivity.java
package com.sysu.edu.academic;

import static com.sysu.edu.api.CommonUtil.extractValue;

import android.os.Bundle;

import com.alibaba.fastjson2.JSONArray;
import com.alibaba.fastjson2.JSONObject;
import com.google.android.material.tabs.TabLayoutMediator;
import com.sysu.edu.BaseActivity;
import com.sysu.edu.R;
import com.sysu.edu.databinding.ActivityPagerBinding;
import com.sysu.edu.model.JwxtModel;
import com.sysu.edu.view.Pager2Adapter;
import com.sysu.edu.view.StaggeredFragment;

import java.util.List;
import java.util.Locale;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class RegistrationActivity extends BaseActivity {
    
    JwxtModel model;
    int page = 0;
    Pager2Adapter adp;
    
    @Override
    protected void onDestroy() {
        super.onDestroy();
        model.dispose();
    }
    
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ActivityPagerBinding binding = ActivityPagerBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        model = new JwxtModel(this);
        adp = new Pager2Adapter(this);
        binding.pager.setAdapter(adp);
        Stream.of("2024", "2025", "2026", "2027").forEach(i -> binding.toolbar.getMenu().add(i).setOnMenuItemClickListener(_ -> {
            ((StaggeredFragment) adp.get(1)).clear();
            getPay(i);
            model.nextAll();
            return false;
        }));
        IntStream.range(0, 3).forEach(i -> {
            adp.add(StaggeredFragment.newInstance(i));
            getNextPage(i);
        });
        new TabLayoutMediator(binding.tabs, binding.pager, (tab, position) -> tab.setText(getResources().getStringArray(R.array.registration_info)[position])).attach();
        binding.toolbar.setNavigationOnClickListener(_ -> supportFinishAfterTransition());
        binding.toolbar.setTitle(R.string.register_info);
        model.getMessage().observe(this, msg -> {
            JSONObject response = msg.getSecond();
            if (response != null && response.getInteger("code").equals(200)) {
                if (response.get("data") != null) {
                    Integer what = msg.getFirst();
                    switch (what) {
                        case 2 -> {
                            JSONObject data = response.getJSONObject("data");
                            int total = data.getInteger("total");
                            data.getJSONArray("rows").forEach(a -> ((StaggeredFragment) adp.get(2)).add(((JSONObject) a).getString("academicYearTerm"), R.drawable.calendar, List.of(new String[]{"学年学期", "校区", "学院", "年级专业", "缴费状态", "报到状态", "注册状态", "报到日期", "注册日期"}),
                                    extractValue((JSONObject) a, new String[]{"academicYearTerm", "campusName", "collegeName", "gradeMajorName", "payedStatusName", "checkInStatusName", "registerStatusName", "checkInDate", "registerDate"})));
                            if (total / 10 > page - 1) getList();
                        }
                        case 0 ->
                                ((StaggeredFragment) adp.get(0)).add("学生报到信息", R.drawable.calendar, List.of("学号", "注册学年学期", "报到状态", "注册状态", "缴费状态"),
                                        extractValue(response.getJSONObject("data"), new String[]{"stuNum", "academicYearTerm", "checkInStatusName", "registerStatusName", "payedStatusName"}));
                        case 1 -> {
                            JSONArray d = response.getJSONArray("data");
                            d.forEach(v -> {
                                StaggeredFragment page2 = (StaggeredFragment) adp.get(1);
                                page2.setHideNull(true);
                                page2.add(((JSONObject) v).getString("acadYear"), R.drawable.money, List.of(new String[]{"年份", "类别", "项目名称", "金额（元）", "区间", "时间"}),
                                        extractValue((JSONObject) v, new String[]{"acadYear", "typeName", "feeTypeName", "payedItemAmount", "feeTimeSection", "editeTime"}));
                            });
                        }
                    }
                    model.nextAll();
                }
            }
        });
        model.next();
    }
    
    void getNextPage(int what) {
        switch (what) {
            case 0 -> getInfo();
            case 1 -> getPay();
            case 2 -> getList();
        }
    }
    
    void getInfo() {
        model.add("jwxt/reports-register/stuRegistration/getSelfRegisterInfo", 0);
    }
    
    void getPay() {
        getPay("2025");
    }
    
    void getPay(String year) {
        model.add("jwxt/reports-register/stuRegistration/getSelfPayInfoDetail?acadYear=" + year, 1);
        
    }
    
    void getList() {
        model.add("jwxt/reports-register/stuRegistration/getSelfRegisterList", String.format(Locale.getDefault(), "{\"pageNo\":%d,\"pageSize\":10,\"total\":true,\"param\":{}}", ++page), 2);
    }
}

File: app/src/main/java/com/sysu/edu/academic/RegistrationActivity.kt
package com.sysu.edu.academic

import android.os.Bundle
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.FilterAlt
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Text
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.mutableStateSetOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.res.stringResource
import androidx.lifecycle.ViewModelProvider
import com.sysu.edu.BaseActivity
import com.sysu.edu.R
import com.sysu.edu.api.CommonUtil.extractValue
import com.sysu.edu.view.ActivityPager
import com.sysu.edu.view.MenuItem
import com.sysu.edu.view.SectionData
import com.sysu.edu.view.StaggerScreen
import com.sysu.edu.view.exportMarkdownMenuItem

class RegistrationActivity : BaseActivity() {
	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		enableEdgeToEdge()
		val viewModel = ViewModelProvider(this)[RegistrationViewModel::class.java]
		
		setContent {
			val registerInfo by viewModel.registerInfo.observeAsState(null)
			val payList by viewModel.payList.observeAsState(emptyList())
			val historyList by viewModel.historyList.observeAsState(emptyList())
			
			LaunchedEffect(Unit) {
				viewModel.fetchRegisterInfo()
				viewModel.fetchPayInfo()
				viewModel.fetchHistoryNextPage()
			}
			val years = remember { mutableStateSetOf<String>() }
			val registerSections = remember(registerInfo) {
				val sections = mutableStateListOf<SectionData>()
				registerInfo?.let { data ->
					sections.add(SectionData(title = getString(R.string.school_enrollment_student_report_info),
					                         icon = R.drawable.calendar,
					                         rows = extractValue(this@RegistrationActivity,
					                                             data,
					                                             intArrayOf(R.string.student_id,
					                                                        R.string.school_enrollment_academic_year,
					                                                        R.string.school_enrollment_checkin_status,
					                                                        R.string.school_enrollment_register_status,
					                                                        R.string.school_enrollment_payment_status),
					                                             arrayOf("stuNum", "academicYearTerm", "checkInStatusName", "registerStatusName", "payedStatusName"))))
				}
				sections
			}
			val paySections = remember(payList) {
				val sections = mutableStateListOf<SectionData>()
				payList.forEach { item ->
					sections.add(SectionData(title = item.getString("acadYear"),
					                         icon = R.drawable.money,
					                         rows = extractValue(this@RegistrationActivity,
					                                             item,
					                                             intArrayOf(R.string.year, R.string.category, R.string.item_name, R.string.amount_yuan, R.string.interval, R.string.time),
					                                             arrayOf("acadYear", "typeName", "feeTypeName", "payedItemAmount", "feeTimeSection", "editeTime"))))
				}
				sections
			}
			val historySections = remember(historyList) {
				val sections = mutableStateListOf<SectionData>()
				years.clear()
				historyList.forEach { item ->
					years.add(item.getString("academicYearTerm").split("-")[0])
					sections.add(SectionData(title = item.getString("academicYearTerm"),
					                         icon = R.drawable.calendar,
					                         rows = extractValue(this@RegistrationActivity,
					                                             item,
					                                             intArrayOf(R.string.school_enrollment_academic_year,
					                                                        R.string.campus,
					                                                        R.string.college,
					                                                        R.string.grade_major,
					                                                        R.string.school_enrollment_payment_status,
					                                                        R.string.school_enrollment_checkin_status,
					                                                        R.string.school_enrollment_register_status,
					                                                        R.string.check_in_date,
					                                                        R.string.register_date),
					                                             arrayOf("academicYearTerm", "campusName", "collegeName", "gradeMajorName", "payedStatusName", "checkInStatusName", "registerStatusName", "checkInDate", "registerDate"))))
				}
				sections
			}
			var dropDownExpanded by remember { mutableStateOf(false) }
			var currentPage by remember { mutableIntStateOf(0) }
			val tabs = listOf(
				MenuItem(stringResource(R.string.registration_current_year)),
				MenuItem(stringResource(R.string.registration_payment)),
				MenuItem(stringResource(R.string.registration_history)),
			                 )
			ActivityPager(title = stringResource(R.string.register_info), tabs = tabs, topBarMenus = { page ->
				mutableListOf(exportMarkdownMenuItem(listOf(registerSections, paySections, historySections), tabs, stringResource(R.string.register_info))).also {
					if (page == 1) it.add(MenuItem(stringResource(R.string.year), Icons.Rounded.FilterAlt,content = {
						DropdownMenu(expanded = dropDownExpanded, onDismissRequest = { dropDownExpanded = false }) {
							years.forEach { year ->
								DropdownMenuItem(text = { Text(year) }, onClick = {
									dropDownExpanded = false
									viewModel.fetchPayInfo(year)
								})
							}
						}
					}) {
						dropDownExpanded = true
						true
					})
				}
			}, onNavigationClick = { supportFinishAfterTransition() }, onPageChange = { currentPage = it }) { page ->
				StaggerScreen(sections = when (page) {
					0 -> registerSections
					1 -> paySections
					2 -> historySections
					else -> registerSections
				}, isHideNull = page == 1, onScrollBottom = {
					if (page == 2 && viewModel.hasMoreHistory()) {
						viewModel.fetchHistoryNextPage()
					}
				})
			}
		}
	}
}